### DISCUSSION TITLE
File Reader Not Reading When Auto Dropped

### DISCUSSION THREAD

**Discussion by IDAccessGoddess** [nextgenhealthcare/connect#5054] [Q&A] [answered]
I have two file reader channels. The source is set up the exact same way for each channel. Same drop directory and move-to directory.
One file is for ADT messages, and the other is for Note messages. They both have different filename patterns.
They both get auto dropped in the SFTP file location by the same user, in the same folder, with the same group permissions. 
The Notes file will drop into the input folder automatically and process successfully, afterwards moving it into the Archive folder as set on the Source tab. 
However, the ADT file will drop into the input folder automatically, does NOT process, but moves it into the Archive folder as set in the Source tab. 
If I manually move the ADT file back into the input folder, it will then process successfully and move back to the Archive folder. I have the same permissions in Unix as the SFTP drop account. 
We have checked permissions and compared the channels, but cannot find the source of why it will not process the ADT file automatically. 
This does not happen in our Test environment, it only happens in the Production environment so I cannot post any screenshots or proprietary coding. 

So, the file does process as it should since it still moves it to the Archive folder, but will only read the file successfully if I re-drop it. 
What are your suggestions for further troubleshooting this issue?

**Comment by pacmano1** [comment]
At least for me, I think you need to clarify your post a bit more.  I don't understand the question because you are mixing terms like "drop", SFTP, input folder, "SFTP drop account", "they", "it", etc.

For example, I might expect something like this:

Mirth v3.12, Windows 2019, OpenJDK 17.

I have two channels, both are file readers that read from the same file system with different file name patterns.  This file system is NFS mounted to the Mirth Server and the mirth server process owner has full permissions to the file system.   

Channel 1 reads ADT with a pattern of "*.hl7"

Channel 2 reads Notes (whatever that means) with a regex pattern of "^N.*.txt" (i.e. the regular expression checkbox is checked)

.. and then whatever your problems is which sounds like "when channel 1 polls it reads an ADT file fine but does not process it, instead moves it to the "after processing folder" . However if I put the same file back in the source directory, it does process it".   I assume here that "process" means that you see the message traverse the source and destinations as expected in the message viewer.

You wrote "can't share code" and "can't share screenshots".    That makes it a bit hard to help.

**Comment by jonbartels** [comment]
Do the filename patterns overlap? Eg is your notes reader picking up your ADTs?

Is there logic in the channel such as a filter or call to `destinationSet.remove` which would skip processing?

Is there any evidence that the misbehaving channel is getting a message at all? This should be message data in the message view from the dashboard or even basic message stats (Eg is the bad channel sitting at zero processed). I can't think of a way that MC would move a file from the read directory to the move-to directory without having a message processed.

What do you see from the MC dashboard on the "connections" tab?

**Reply by IDAccessGoddess** [reply]
No, they don't overlap. The names are completely different. 

I did not find any logic that would skip processing. 

I know the message has to be processing somehow through MC because it's moving it to the Archive folder. If it wasn't processing the file, it would just stay stuck in the input folder, right? 

Unfortunately, the connection log doesn't go that far back do to the amount of channels we have and I'll have to catch it tomorrow since the files drop at 5:30 PM daily. 

Filename for the Note: EpicOpptyDispNotes_20220223220024.txt
Filename Filter Pattern: EpicOppty*

Filename for ADT: EPIC_MBR_DEMO_20220223.txt
Filename Filter Pattern: EPIC_MBR_DEMO_*

**Reply by jonbartels** [reply]
Hmm. Do both file patterns exist in the same folder? Is it possible that there is a channel lurking with a `*.*` pattern thats snaffling up the messages?
I cannot articulate why but I can't help but think that some other channel is picking up the same messages and moving them. Occams Razor I guess.

> I know the message has to be processing somehow through MC because it's moving it to the Archive folder. If it wasn't processing the file, it would just stay stuck in the input folder, right?

This is correct. MC physically has to process a message and the **source connector** has to be in a terminal state before it will move the file.

Slack may be a good option to talk to gurus on the fly https://mirthconnect.herokuapp.com/ . Paul is correct that sharing or seeing some details is important to get good help. Slack may be a good way to share some details (never sharing PHI nor credentials of course!) in a transient way instead of leaving screenshots on github for all time.

The "Free 15" from my employer may also be a good way to show some details with reasonable protections for any secret-sauce, https://consultzen.com/free-interoperability-expert-consult/

**Reply by IDAccessGoddess** [reply]
Okay, so I got screenshots of the connection log on the channel. The first time the file drops, it will Read the file, but not Write. And once I re-drop the file it will Read AND start Writing. 

![image](https://user-images.githubusercontent.com/88687189/155813828-ab045bcc-8290-4e21-86f6-bfbba9705b7e.png)

![image](https://user-images.githubusercontent.com/88687189/155813869-14826965-fc21-4c96-8c0c-4b39f450c8c6.png)

**Reply by jonbartels** [reply]
The screenshots help visualize the problem, thank you.

I assume for each operation the "Channel" is the same? 

Based on the first drop not reading, increase your "Check File Age" setting. If that is tool low MC could read the file before its totally written. However, the symptoms don't jive. I'd expect that to error or transmit a partial file.

**Reply by IDAccessGoddess** [reply]
Yes. It's the same channel. 
And we have the Check File Age set to No. Should we turn it on to Yes? 

![image](https://user-images.githubusercontent.com/88687189/156006139-a5c474f8-abff-4d21-bb2a-bc33d7e6402d.png)

**Reply by pacmano1** [reply]
That's always a good idea and typically setting a minimum size is as well.

**Reply by IDAccessGoddess** [reply]
What would you suggest on the File Age requirements?

**Reply by pacmano1** [reply]
Most of my stuff is 10 seconds.  Depends on your use case and file sizes written but that has seems to work well.  That time is in UI is specified in milliseconds of course.  I almost always set a 100 byte file size minimum but that may not be relevant in your use case.

**Reply by IDAccessGoddess** [reply]
These files are usually 7MB.

**Reply by jonbartels** [reply]
Rough suggestions.

For a daily file drop of large XML or delimited text - Set the size relatively large, you want it small enough to catch even the smallest files dropped on a Saturday over the winter holidays but big enough to not prematurely read a file. Then a long file age value is fine, if files are dropping every 24 hours it's not like a few minutes delay will matter since its a slow asynchronous process anyway.

For a use case that requires fast responses - First a poller is NOT the right tool to get fast response times! If you have to use it then tuning file age and file size right on the edge of the acceptable responsiveness is needed. This really is not a good pattern because if you set the size too large you miss small drops; if you set the file age too high, you murder your overall processing time just idling.

**Reply by pacmano1** [reply]
I might wait a minute or two.   I think you can compare the created date to last modified date on some file systems to determine how long files are taking to write.   Then wait longer than the max diff.

**Reply by IDAccessGoddess** [reply]
Thanks! Let me try this and I'll see how it goes. You all are my heros for the day. :)

**Reply by IDAccessGoddess** [reply]
So, it ended up being a combination of the Check File Age and The Polling Interval. Thanks for all your help.

**Comment by IDAccessGoddess** [comment, marked as answer]
So, it ended up being a combination of the Check File Age and The Polling Interval. Thanks for all your help.
